File: 04-polimorfismo/java/Demo04.java
/*
 * DEMO 04 - POLIMORFISMO, CLASES ABSTRACTAS E INTERFACES
 *
 * Compilar y ejecutar:
 *     javac *.java
 *     java Demo04
 *
 * Seis archivos:
 *     Figura.java      la idea abstracta
 *     Dibujable.java   el contrato
 *     Circulo.java     Figura + Dibujable
 *     Rectangulo.java  Figura + Dibujable
 *     Triangulo.java   solo Figura
 *     Demo04.java      el programa   <- estas aqui
 *
 * Idea clave: podemos tratar a muchos objetos distintos COMO SI FUERAN
 * el mismo tipo, y cada uno responde a su manera.
 * "Un mismo llamado, muchas respuestas."
 */
public class Demo04 {
    public static void main(String[] args) {

        // Un arreglo de Figura que en realidad guarda objetos de 3 clases distintas
        Figura[] figuras = {
            new Circulo(3),
            new Rectangulo(4, 5),
            new Triangulo(6, 2)
        };

        System.out.println("--- POLIMORFISMO: el mismo llamado, distinta respuesta ---");
        for (Figura f : figuras) {
            // Java decide en TIEMPO DE EJECUCION cual area() ejecutar
            System.out.printf("%-12s area = %6.2f | perimetro = %6.2f%n",
                              f.getNombre(), f.area(), f.perimetro());
        }

        System.out.println("--- Un metodo que sirve para cualquier Figura ---");
        System.out.printf("Area total: %.2f%n", sumarAreas(figuras));

        System.out.println("--- INTERFAZ: un contrato aparte de la herencia ---");
        for (Figura f : figuras) {
            if (f instanceof Dibujable) {          // ¿cumple el contrato?
                ((Dibujable) f).dibujar();
            } else {
                System.out.println(f.getNombre() + " no sabe dibujarse.");
            }
        }
    }

    // Este metodo NO sabe ni le importa que figuras concretas le pasen.
    // Solo sabe que todas responden a area(). Eso es programar contra la abstraccion.
    static double sumarAreas(Figura[] figuras) {
        double total = 0;
        for (Figura f : figuras) {
            total += f.area();
        }
        return total;
    }
}

File: 04-polimorfismo/python/demo04.py
"""
DEMO 04 - POLIMORFISMO, CLASES ABSTRACTAS E INTERFACES
Ejecutar:  python3 demo04.py

Siete archivos:
    figura.py      la idea abstracta
    dibujable.py   el contrato
    circulo.py     Figura + Dibujable
    rectangulo.py  Figura + Dibujable
    triangulo.py   solo Figura
    cuadrado.py    NADA (duck typing)
    demo04.py      el programa   <- estas aqui

  f instanceof Dibujable  ->  isinstance(f, Dibujable)
"""

from circulo import Circulo
from cuadrado import Cuadrado
from dibujable import Dibujable
from figura import Figura
from rectangulo import Rectangulo
from triangulo import Triangulo


def sumar_areas(figuras):
    """No sabe ni le importa que figuras le pasen. Solo que responden a area()."""
    return sum(f.area() for f in figuras)


if __name__ == "__main__":

    figuras = [Circulo(3.0), Rectangulo(4.0, 5.0), Triangulo(6.0, 2.0)]

    print("--- POLIMORFISMO: el mismo llamado, distinta respuesta ---")
    for f in figuras:
        print(f"{f.get_nombre():<12} area = {f.area():6.2f} | perimetro = {f.perimetro():6.2f}")

    print("--- Un metodo que sirve para cualquier Figura ---")
    print(f"Area total: {sumar_areas(figuras):.2f}")

    print("--- INTERFAZ: un contrato aparte de la herencia ---")
    for f in figuras:
        if isinstance(f, Dibujable):
            f.dibujar()
        else:
            print(f"{f.get_nombre()} no sabe dibujarse.")

    print("--- La clase abstracta no se puede instanciar ---")
    try:
        f = Figura("fantasma")
    except TypeError as e:
        print(f"Error esperado -> {e}")

    print("--- DUCK TYPING: lo que Java NO permite ---")
    patito = Cuadrado()          # no hereda de Figura. De nada.
    print(f"{patito.get_nombre():<12} area = {patito.area():6.2f}")
    print(f"Area total con el intruso: {sumar_areas(figuras + [patito]):.2f}")
    print("Funciono sin heredar de Figura. En Java esto NO compila.")
